File: uf.py
def union(x, y):
	 xRoot = find(x)
	 yRoot = find(y)
	 if xRoot.rank > yRoot.rank:
		 yRoot.parent = xRoot
	 elif xRoot.rank < yRoot.rank:
		 xRoot.parent = yRoot
	 elif xRoot != yRoot: # Unless x and y are already in same set, merge them
		 yRoot.parent = xRoot
		 xRoot.rank = xRoot.rank + 1

def find(x):
	 if x.parent == x:
		return x
	 else:
		x.parent = find(x.parent)
		return x.parent

File: uf.c
#include "uf.h"

void make_set(Vertex *v){
	v->parent = v ;
	v->rank = 0;
}
Vertex *find(Vertex *v){
	if(v->parent == v){
		return v;
	}else{
		return	v->parent = find(v->parent) ;
	}
}
void merge(Vertex *x, Vertex *y){
	Vertex *x_root = find(x) ;
	Vertex *y_root = find(y) ;
	if(x_root->rank > y_root->rank){
		y_root->parent = x_root ;
	}else if(x_root->rank < y_root->rank){
		x_root->parent = y_root ;
	}else if(x_root != y_root){
		y_root->parent = x_root ;
		x_root->rank = x_root->rank + 1 ;
	}
}

